text/0000-execution-environments.md

```diff
@@ -0,0 +1,225 @@
+# Meta
+[meta]: #meta
+- Name: Execution Environments
+- Start Date: 2023-01-11
+- Author(s): @hone
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+Add support for different execution environments for buildpacks, with testing as the first use case.
+
+# Definitions
+[definitions]: #definitions
+
+- Execution Environment - The target environment the OCI image is expected to be run in, i.e. production, test, development.
+
+# Motivation
+[motivation]: #motivation
+
+The main way Buildpacks are being used is building production images, but this is only one piece of the CI/CD pipeline. Without a solid buildpack test environment story, users will be required alternatives for building their environment. A testing environment, while different, shares many of the same broad strokes ultimately producing an execution environment. This is something Buildpacks are well suited to solve.
+
+# What it is
+[what-it-is]: #what-it-is
+
+## Test Environments
+
+One of the goals in this design is to minimize the changes needed. The Buildpack lifecycle is already well suited to produce environments:
+
+```
+App Source -> OCI Image (Production Environment) -> Execute (Launch Process)
+```
+
+To create a test environment, it can follow the same flow:
+
+```
+App Source -> OCI Image (Test Environment) -> Execute (Test Process)
+```
+
+### Division of Responsibility
+With the test OCI Image, a platform can execute the tests in the pipeline as they see fit. This means a bulk of the responsibilities are platform concerns:
+
+- Set which environment to build for
+- Decide which buildpacks to execute
+- How to execute the tests
+- What is the test result format like [TAP](https://en.wikipedia.org/wiki/Test_Anything_Protocol)?
+- How to process the test results
+- What to do with the results
+
+This narrows the scope of the Buildpack responsibilities to what it's already good at:
+
+- Produce Test Environment
+- How to launch the process
+  - launch.toml with test process marked as default (recommendation)
+
+## Setting the Execution Enviroment for Build + Buildpack
+A platform will set the `CNB_EXEC_ENV` env var to the execution environment desired. Buildpacks can than read this env var to branch or switch on logic needed based on the execution environment.
+
+In addition, Builder Authors, Buildpack Authors, and App Developers will be able to specify various options to a specific execution enviroment using the `exec-env` key.
+
+## Development Environments
+The specifics of creating development enviroments are out of scope of this RFC, but it's not hard to extrapolate how these kind of changes can assist in creating Buildpacks for development environments.
+
+# How it Works
```

Comment:
It would be nice to see the pack flow as well

Reply:
Should we add anything to the app image to designate it as being built for a particular environment? To avoid users accidentally deploying a test image in production...

I could see folks wanting to use the same tag when re-building a test image for production, in order to use previously cached dependencies.

Reply:
@natalieparellano that's a good question. Do you think cached dependencies should be shared b/t different execution environments?

Reply:
Is a label the best place to designate the execution environment?